File: platform-core/platform-cache/src/main/scala/org/sunbird/cache/mgr/LocalCacheManager.scala
package org.sunbird.cache.mgr

import java.util.concurrent.Executors

import org.sunbird.cache.common.CacheErrorCode
import org.sunbird.cache.util.RedisCache
import org.sunbird.common.exception.ServerException
import org.sunbird.telemetry.logger.TelemetryManager
import redis.clients.jedis.JedisPubSub

/**
 * Abstract Implementation For Local Cache
 * @see ICacheManager
 * @tparam T
 */
abstract class LocalCacheManager[T] extends ICacheManager[T] {

	protected def init(channels: String*): Unit = {
		subscribe(channels.map(_.asInstanceOf[String]): _*)
	}

	override def publish(channel: String, message: String): Unit = {
		RedisCache.publish(channel, message)
	}

	override def subscribe(channels: String*): Unit = {
		val pubSub = new JedisPubSub() {
			override def onMessage(channel: String, message: String): Unit = {
				processSubscription(channel, message)
			}
		}
		val pool = Executors.newFixedThreadPool(1)
		try {
			pool.execute(new Runnable() {
				override def run(): Unit = {
					RedisCache.subscribe(pubSub, channels.map(_.asInstanceOf[String]): _*)
				}
			})
		} catch {
			case e: Exception =>
				TelemetryManager.error("Exception Occurred While Subscribing to channels : " + channels + " | Exception is : " + e)
				throw new ServerException(CacheErrorCode.ERR_CACHE_SUBSCRIBE_CHANNEL_ERROR.toString, e.getMessage)
		} finally if (null != pool) pool.shutdown()
	}

	protected def processSubscription(channel: String, message: String): Unit

}

File: platform-core/platform-cache/src/main/java/org/sunbird/cache/mgr/LocalCacheManager.java
package org.sunbird.cache.mgr;

import org.sunbird.cache.common.CacheErrorCode;
import org.sunbird.cache.handler.ICacheHandler;
import org.sunbird.cache.util.RedisCacheUtil;
import org.sunbird.common.exception.ServerException;
import org.sunbird.telemetry.logger.TelemetryManager;
import redis.clients.jedis.JedisPubSub;

import java.util.List;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;

public abstract class LocalCacheManager implements ICacheManager {

    ICacheHandler handler;

    protected void init(String... channels){
        subscribe(channels);
    }

    @Override
    public void publish(String channel, String message) {
        RedisCacheUtil.publish(channel, message);
    }

    @Override
    public void subscribe(String... channels) {
        JedisPubSub pubSub = new JedisPubSub() {
            @Override
            public void onMessage(String channel, String message) {
                processSubscription(channel, message);
            }
        };

        ExecutorService pool = null;
        try {
            pool = Executors.newFixedThreadPool(1);
            pool.execute(new Runnable() {
                public void run() {
                    RedisCacheUtil.subscribe(pubSub, channels);
                }
            });
        } catch (Exception e) {
            TelemetryManager.error("Exception Occured While Subscribing to channels : " + channels + " | Exception is : " + e);
            throw new ServerException(CacheErrorCode.ERR_CACHE_SUBSCRIBE_CHANNEL_ERROR.name(),e.getMessage());
        } finally {
            if (null != pool)
                pool.shutdown();
        }
    }

    protected abstract void processSubscription(String channel, String message);


    @Override
    public String getString(String key) {
        return null;
    }

    @Override
    public void setString(String key, String data, int ttl) {

    }

    @Override
    public List<String> getList(String key) {
        return null;
    }

    @Override
    public void setList(String key, List<String> list, int ttl) {

    }

    @Override
    public void increment(String key) {

    }

}
